File: 프로그래머스/level3/경주로 건설.js
let dr = [-1, 0, 1, 0];
let dc = [0, 1, 0, -1];

function solution(board) {
    var answer = 0;
    let visit = [];
    let N = board.length;
    for(let i=0; i<board.length; i++){
        let tmp = new Array(board[i].length).fill(987654321);
        visit.push(tmp);
    }
    let q = [];
    q.push([0, 0, -1, 0]);
    visit[0][0] = 0;
    while(q.length>0){
        let r = q[0][0];
        let c = q[0][1];
        let di = q[0][2];
        let cost = q[0][3];
        q.shift();
        for(let d=0; d<4; d++){
            if(di!==-1&&d===(di+2)%4)    continue;
            let nr = r + dr[d];
            let nc = c + dc[d];
            let ncost = cost + 100;
            if(di!==-1&&d!==di)  ncost += 500;
            if(nr<0||nc<0||nr>=N||nc>=N||board[nr][nc]===1)    continue;
            if(visit[nr][nc]<ncost)         continue;
            visit[nr][nc] = ncost;
            q.push([nr, nc, d, ncost]);
        }
    }
    answer = visit[N-1][N-1];
    return answer;
}

File: 프로그래머스/level3/경주로 건설.cpp
//BFS
#include <string>
#include <tuple>
#include <vector>
#include <queue>

#define tiii tuple<int, int, int, int>

using namespace std;

int dr[4] = {-1, 0, 1, 0};
int dc[4] = {0, 1, 0, -1};

int cost[26][26];
int N;

int solution(vector<vector<int>> board) {
    int answer = 0;
    N = board.size();
    for(int i=0; i<N; i++){
        for(int j=0; j<N; j++){
            cost[i][j] = 0x3f3f3f3f;
        }
    }
    cost[0][0] = 0;
    queue<tiii> q;
    q.push({0,0,1,0});
    q.push({0,0,2,0});
    while(!q.empty()){
        auto [r, c, d, now] = q.front();
        q.pop();
        for(int nd=0; nd<4; nd++){
            int nr = r + dr[nd];
            int nc = c + dc[nd];
            if(nr<0||nc<0||nr>=N||nc>=N)    continue;
            if(board[nr][nc] == 1)          continue;
            int plus = 100;
            if((nd+1)%4 == d || (d+1)%4 == nd)  plus += 500;
            int next = now + plus;
            if(cost[nr][nc] + 600 < next)         continue;
            if(cost[nr][nc]>next)   cost[nr][nc] = next;
            q.push({nr, nc, nd, next});
        }
    }
    answer = cost[N-1][N-1];
    return answer;
}